File: github_code/shell-sort/cpp/117.c
/*
 * Assignment #12: Compare sorting algorithms
 *
 * CMPE 180-92 Data Structures and Algorithms in C++
 * Department of Computer Engineering
 * R. Mak, Nov. 20, 2016
 */
#include "ShellSortSuboptimal.h"

/**
 * Default constructor.
 */
ShellSortSuboptimal::ShellSortSuboptimal() {}

/**
 * Destructor.
 */
ShellSortSuboptimal::~ShellSortSuboptimal() {}

/**
 * Get the name of this sorting algorithm.
 * @return the name.
 */
string ShellSortSuboptimal::name() const { return "Shellsort suboptimal"; }

/**
 * Run the suboptimal shellsort algorithm.
 * @throws an exception if an error occurred.
 */
void ShellSortSuboptimal::run_sort_algorithm() throw (string)
{
    /***** Complete this member function. *****/
	//source::http://www.programming-algorithms.net/article/41653/Shell-sort
	int g = size / 2;
while (g > 0) {
for (int i = 0; i < size - g; i++) { 
int j = i + g;
int tmp = array[j];
while (j >= g && tmp > array[j - g]) {
array[j] = array[j - g];
j -= g;
}
array[j] = tmp;
}
if (g == 2) { 

g = 1;

} else {

g =g/2;

}

}
return array;
}
}

File: github_code_index/java/13/117.java
package sorting;

public class ShellSort {
    public static int ITERATION_COEFFICIENT = 2;

    public static void shellSort(int[] data) {
        for (int step = data.length / ITERATION_COEFFICIENT; step >= 1; step /= ITERATION_COEFFICIENT) {
            for (int i = 0; i < step; i++) {
                for (int j = i + step; j < data.length; j += step) {
                    if (data[j] < data[j - step]) {
                        int tmp = data[j];
                        int k = j - step;
                        for (; k >= 0 && data[k] > tmp; k -= step) {
                            data[k + step] = data[k];
                        }
                        data[k + step] = tmp;
                    }
                }
            }
        }
    }

    public static void shellSortOptim(int[] data) {
        int step = data.length / ITERATION_COEFFICIENT;
        for (; step >= 1; step /= ITERATION_COEFFICIENT) {
            for (int i = step; i < data.length; i++) {
                int tmp = data[i];
                int k = i - step;
                if (data[i] >= data[k]) {
                    continue;
                }
                for (; k >= 0 && data[k] > tmp; k -= step) {
                    data[k + step] = data[k];
                }
                data[k + step] = tmp;
            }
        }

    }
}
